File: dsa_with_python/linked_list/move_zeroes_to_front.py
from node import Node
from singly_linked_list import SinglyLinkedList


def move_zeroes(head: Node) -> Node:
    prev_node = head
    current_node = head.next
    while current_node is not None:
        if int(current_node.data) == 0:
            prev_node.next = current_node.next
            current_node.next = head
            head = current_node
            current_node = prev_node.next
            continue
        prev_node = prev_node.next
        current_node = current_node.next
    return head


if __name__ == "__main__":
    sll = SinglyLinkedList()
    while True:
        try:
            choice = int(
                input(
                    "Press:\n"
                    "1 to Insert\n"
                    "2 to Display\n"
                    "3 to Delete All\n"
                    "4 to Move Zeroes to Front\n"
                    "5 to exit()\n"
                    "Enter your choice -> "
                )
            )
        except Exception:
            print(f"Wrong Input! Only Integer value is allowed.")
            continue
        match choice:
            case 1:
                data = input("Enter the Data -> ")
                sll.insert_at_last(data)
            case 2:
                sll.display()
            case 3:
                sll.delete_all()
            case 4:
                if not sll.is_empty():
                    sll.head = move_zeroes(sll.head)
                else:
                    print("List is Empty!")
            case 5:
                exit(0)
            case _:
                print("Wrong Input!")

File: Java/linkedlist/MoveZeroesToFront.java
package linkedlist;

import java.util.Scanner;

import definitions.ListNode;

public class MoveZeroesToFront extends SinglyLinkedList {
    static Scanner sc = new Scanner(System.in);

    public ListNode moveZeroes(ListNode head) {
        ListNode prevNode = head;
        ListNode currentNode = head.next;
        while (currentNode != null) {
            if (currentNode.data == 0) {
                prevNode.next = currentNode.next;
                currentNode.next = head;
                head = currentNode;
                currentNode = prevNode.next;
                continue;
            }
            prevNode = prevNode.next;
            currentNode = currentNode.next;
        }
        return head;
    }

    public static void main(String[] args) {
        MoveZeroesToFront ob = new MoveZeroesToFront();
        while (true) {
            int choice, data;
            System.out.print("Press:\n" +
                    "1 to Insert\n" +
                    "2 to Display\n" +
                    "3 to Delete All\n" +
                    "4 to Move Zeroes to Front\n" +
                    "5 to exit()\n" +
                    "Enter your choice -> ");
            try {
                choice = sc.nextInt();
            } catch (Exception e) {
                System.out.println("Wrong Input! Only Integer value is allowed.");
                sc.next();
                continue;
            }
            switch (choice) {
                case 1 -> {
                    System.out.print("Enter your Data -> ");
                    data = sc.nextInt();
                    ob.insertAtLast(data);
                }
                case 2 ->
                    ob.display();
                case 3 ->
                    ob.deleteAll();
                case 4 -> {
                    if (!ob.isEmpty())
                        ob.head = ob.moveZeroes(ob.head);
                    else
                        System.out.println("List is Empty!");
                }
                case 5 ->
                    System.exit(0);
                default ->
                    System.out.println("Wrong Choice!");
            }
        }
    }
}
